ARROW-13224: [Python][Doc] Documentation missing for pyarrow.dataset.write_dataset (#10693)
apache/arrow: docs/source/python/dataset.rst

    dataset.to_table().to_pandas()
    dataset.to_table(filter=ds.field('year') == 2019).to_pandas()

> Another benefit of manually scheduling the files is that the order of the files

Reviewer: "Scheduling" is a bit strange term to use in this context?

Author: Yes it is.  I changed it to "listing"

controls the order of the data.  When performing an ordered read (or a read to
a table) then the rows returned will match the order of the files given.

Iterative (out of core or streaming) reads
------------------------------------------

> The previous examples have demonstrated how to read the data into a table.  This is

Reviewer: ```suggestion
The previous examples have demonstrated how to read the data into a table using :func:`~Dataset.to_table`.  This is
```

?

useful if the dataset is small or there is only a small amount of data that needs to
be read.  The dataset API contains additional methods to read and process large amounts
of data in a streaming fashion.

The easiest way to do this is to use the method :meth:`Dataset.to_batches`.  This
method returns an iterator of record batches.  For example, we can use this method to
calculate the average of a column without loading the entire column into memory:

.. ipython:: python

    import pyarrow.compute as pc

    col2_sum = 0
    count = 0

>     for batch in dataset.to_batches(filter=~ds.field('col2').is_null()):

Reviewer: Maybe add a `columns=["col2"]` as well? (it shows passing a column selection as well + it is actually what you should do here if you are only summing that column)

Author: Added

        col2_sum += pc.sum(batch.column('col2')).as_py()
        count += batch.num_rows
    mean_a = col2_sum/count

Customizing the batch size
~~~~~~~~~~~~~~~~~~~~~~~~~~

An iterative read of a dataset is often called a "scan" of the dataset and pyarrow
uses an object called a :class:`Scanner` to do this.  A Scanner is created for you
automatically by the to_table and to_batches method of the dataset.  Any arguments
you pass to these methods will be passed on to the Scanner constructor.

> One of those parameters is the `batch_size`.  This controls the maximum size of the

Reviewer: ```suggestion
One of those parameters is the ``batch_size``.  This controls the maximum size of the
```

(rst uses double backticks for code, in contrast to markdown ..)

batches returned by the scanner.  Batches can still be smaller than the `batch_size`
if the dataset consists of small files or those files themselves consist of small
row groups.  For example, a parquet file with 10,000 rows per row group will yield
batches with, at most, 10,000 rows unless the batch_size is set to a smaller value.

The default batch size is one million rows and this is typically a good default but
you may want to customize it if you are reading a large number of columns.

Writing Datasets
----------------

> The dataset API also simplifies writing data to a dataset.  This can be useful when

Reviewer: ```suggestion
The dataset API also simplifies writing data to a dataset using :func:`write_dataset` .  This can be useful when
```

(this is not necessarily the best addition, but basically I think it would be good to add this explicit reference _somewhere_ in this paragraph; it's useful to have this so that the user has an easy link to click to check the more detailed function docstring)

you may not be able to load everything into a single in-memory table.  Fortunately, the
write_dataset method also accepts an iterable of record batches.  This makes it really
simple, for example, to repartition a large dataset without loading the entire dataset
into memory:

.. ipython:: python

    old_part = ds.partitioning(
        pa.schema([("c", pa.int16())]), flavor="hive"
    )
    new_part = ds.partitioning(
        pa.schema([("c", pa.int16())]), flavor=None
    )
    input_dataset = ds.dataset(dataset_root, partitioning=old_part)
    new_root = base / "repartitioned_dataset"
    # A scanner can act as an iterator of record batches but you could also receive
    # data from the network (e.g. via flight), from your own scanning, or from any
    # other method that yields record batches.  In addition, you can pass a dataset
    # into write_dataset directly but this method is useful if you want to customize
    # the scanner (e.g. to filter the input dataset or set a maximum batch size)
    scanner = input_dataset.scanner()

    ds.write_dataset(scanner, new_root, format="parquet", partitioning=new_part)

> After the above example runs our data will be in dataset_root/1 and dataset_root/2

Reviewer: Since this example is not changing the partitioning structure (except for the flavor), it's not directly what this "repartitioning" will do apart from changing `/c=1/` into `/1/`. Will it also repartition individual files? (eg write more smaller files in case your input dataset has large files)

Author: This would only change the directory structure.  I didn't want to make the example data more complicated and `a` and `b` would be bad choices for a partition column.  I modified the following paragraph to hopefully make it a bit clearer...

> In this simple example we are not changing the structure of the data
> (only the directory naming schema) but you could also use this mechnaism to change
> which columns are used to partition the dataset.  This is useful when you expect to
> query your data in specific ways and you can utilize partitioning to reduce the
> amount of data you need to read.

> directories.  You could also use this mechnaism to change which columns you are partitioned

Reviewer: ```suggestion
directories.  You could also use this mechnaism to change which columns the dataset is partitioned
```

Author: I changed to `you could also use this mechnaism to change which columns are used to partition the dataset`

    dataset.to_table().to_pandas()
    dataset.to_table(filter=ds.field('year') == 2019).to_pandas()

Another benefit of manually scheduling the files is that the order of the files
controls the order of the data.  When performing an ordered read (or a read to
a table) then the rows returned will match the order of the files given.  This
only applies when the dataset is constructed with a list of files.  There

> are no order guarantees given when the files are instead discovered by scanning

Reviewer 2: We do sort by path: 
https://github.com/apache/arrow/blob/91f261fa9a7841fd914c5ed1d8e747fb4e510a5b/cpp/src/arrow/dataset/discovery.cc#L203-L204

Though it states that's only for unit tests. However I would personally be really surprised if data were truly returned in random order and as we've had this behavior in a few releases now I think users would also be surprised. Maybe we should instead guarantee lexicographic sort on paths?

Author: Hmmm, I just dropped the sentence about scanning.  If someone wants clarification we can discuss it at that time.  @bkietz any opinion?  I recall we spoke about this before.

Reviewer 2: I also suppose we don't have to solve this here; we can get the docs in and improve this part later.

Reviewer 3: We don't guarantee order for selectors because ARROW-8163 (asynchronous fragment discovery) might not guarantee order. Lexicographic sorting *could* be maintained for synchronous discovery from a selector, but in general we'd want to push a fragment into scan as soon as it's yielded by `FileSystem::GetFileInfoGenerator`

Reviewer 2: Ah thanks. Sorry for the churn here @westonpace - we should keep stating that there's no guarantee then.

a directory.

Iterative (out of core or streaming) reads
------------------------------------------

The previous examples have demonstrated how to read the data into a table.  This is
useful if the dataset is small or there is only a small amount of data that needs to
be read.  The dataset API contains additional methods to read and process large amounts
of data in a streaming fashion.

The easiest way to do this is to use the method :meth:`Dataset.to_batches`.  This
method returns an iterator of record batches.  For example, we can use this method to
calculate the average of a column without loading the entire column into memory:

> .. ipython:: python

Reviewer 2: Do we actually want all these to be ipython blocks instead of code-block blocks? It does slow down building a bit.

Author: All the scripts I added execute pretty rapidly as they are dealing with tables with less than 10 rows.  I'm not sure they add significantly to the build times.

For a test I tried converting all ipython to code-block and saw no noticable difference in build times.  I'd prefer ipython just for the testing sake but I'm happy to go with whatever is decided in ARROW-13159.  Since it should be pretty easy to change after the fact (just a find-replace from ipython to code-block) I'd rather address it after ARROW-13159 is resolved if that is ok.

Reviewer 2: Ah, thanks for checking.

    import pyarrow.compute as pc

>     sum = 0

Reviewer 2: nit: this shadows a builtin

Author: Thanks.  I renamed it.

    count = 0
    for batch in dataset.to_batches(filter=~ds.field('col2').is_null()):
        sum += pc.sum(batch.column('col2')).as_py()
        count += batch.num_rows

>     mean_a = sum/count

Reviewer 2: Though, wouldn't the 'right' way to do it be to select `mean('col2')` as an expression? (I suppose you can't yet do that though.)

Author: True, but then you wouldn't be using `to_batches` anyways and this whole section goes away.  I think dataset scanning is likely to be a niche API once any kind of SQL support is added.  We can address it then.

batches returned by the scanner.  Batches can still be smaller than the `batch_size`
if the dataset consists of small files or those files themselves consist of small
row groups.  For example, a parquet file with 10,000 rows per row group will yield
batches with, at most, 10,000 rows unless the batch_size is set to a smaller value.

The default batch size is one million rows and this is typically a good default but
you may want to customize it if you are reading a large number of columns.

Writing Datasets
----------------

The dataset API also simplifies writing data to a dataset.  This can be useful when
you want to partition your data or you need to write a large amount of data.  A
basic dataset write is similar to writing a table except that you specify a directory
instead of a filename.

.. ipython:: python

    base = pathlib.Path(tempfile.gettempdir())
    dataset_root = base / "sample_dataset"
    dataset_root.mkdir(exist_ok=True)

    table = pa.table({"a": range(10), "b": np.random.randn(10), "c": [1, 2] * 5})
    ds.write_dataset(table, dataset_root, format="parquet")

The above example will create a single file named part-0.parquet in our sample_dataset

> diretory.

Reviewer 2: ```suggestion
directory.
```

Reviewer 2: Is it worth noting IPC (Feather) and CSV are also supported?

    larger or partitioned datasets it is not so easy.  The ``file_visitor`` keyword can be used 
    to supply a visitor that will be called as each file is created:

    .. ipython:: python

        def file_visitor(written_file):
            print(f"path={written_file.path}")
            print(f"metadata={written_file.metadata}")
        ds.write_dataset(table, dataset_root, format="parquet", partitioning=part,
                        file_visitor=file_visitor)

    This will allow you to collect the filenames that belong to the dataset and store them elsewhere
    which can be useful when you want to avoid scanning directories the next time you need to read
    the data.  It can also be used to generate the _metadata index file used by other tools such as
    dask or spark to create an index of the dataset.

Configuring format-specific parameters during a write
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In addition to the common options shared by all formats there are also format specific options
that are unique to a particular format.  For example, to allow truncated timestamps while writing

> parquet files:

Reviewer 2: ```suggestion
Parquet files:
```

Reviewer 2: Is it worth linking to the options classes for each of the formats?

row groups.  For example, a parquet file with 10,000 rows per row group will yield
batches with, at most, 10,000 rows unless the batch_size is set to a smaller value.

The default batch size is one million rows and this is typically a good default but
you may want to customize it if you are reading a large number of columns.

Writing Datasets
----------------

The dataset API also simplifies writing data to a dataset using :func:`write_dataset` .  This can be useful when
you want to partition your data or you need to write a large amount of data.  A
basic dataset write is similar to writing a table except that you specify a directory
instead of a filename.

.. ipython:: python

    base = pathlib.Path(tempfile.gettempdir())
    dataset_root = base / "sample_dataset"
    dataset_root.mkdir(exist_ok=True)

    table = pa.table({"a": range(10), "b": np.random.randn(10), "c": [1, 2] * 5})
    ds.write_dataset(table, dataset_root, format="parquet")

The above example will create a single file named part-0.parquet in our sample_dataset
directory.

.. warning::

    If you run the example again it will replace the existing part-0.parquet file.

>     Appending files to an existing dataset is not currently supported by this API and
>     the output directory should be empty for predictable results.

Reviewer 3: ```suggestion
    Appending files to an existing dataset requires specifying a new
    ``basename_template`` for each call to ``ds.write_dataset``
    to avoid overwrite.
```
